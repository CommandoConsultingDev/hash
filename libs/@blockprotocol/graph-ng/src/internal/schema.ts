export type UndefinedOnPartialShallow<T> = T extends Record<any, any>
  ? {
      [Key in keyof T]: undefined extends T[Key] ? T[Key] | undefined : T[Key];
    }
  : T;

export type PartialOnUndefinedShallow<T> = T extends Record<any, any>
  ? {
      [Key in keyof T as undefined extends T[Key]
        ? Key
        : never]?: T[Key] extends infer U | undefined ? U : never;
    } & {
      [Key in keyof T as undefined extends T[Key] ? never : Key]: T[Key];
    }
  : T;

export function pruneUndefinedShallow<T extends Record<any, any>>(
  value: T,
): PartialOnUndefinedShallow<T> {
  const result: any = {};

  for (const key in value) {
    if (value[key] !== undefined) {
      result[key] = value[key];
    }
  }

  return result;
}

// Taken from:
// <https://github.com/sindresorhus/escape-string-regexp/blob/ba9a4473850cb367936417e97f1f2191b7cc67dd/index.js>
// (Licensed under MIT)
export function escapeStringRegexp(value: string) {
  // Escape characters with special meaning either inside or outside character sets.
  // Use a simple backslash escape when it’s always valid, and a `\xnn` escape when the simpler form would be disallowed by Unicode patterns’ stricter grammar.
  return value.replace(/[|\\{}()[\]^$+*?.]/g, "\\$&").replace(/-/g, "\\x2d");
}
