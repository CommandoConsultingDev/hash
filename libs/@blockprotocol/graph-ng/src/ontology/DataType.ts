import * as S from "@effect/schema/Schema";
import {
  Effect,
  Option,
  Equal,
  Hash,
  Inspectable,
  pipe,
  Pipeable,
  Predicate,
} from "effect";
import { globalValue } from "effect/GlobalValue";

import * as Json from "../Json.js";
import { decodeSchema } from "./DataType/decode.js";
import { encodeSchema } from "./DataType/encode.js";
import { DecodeError, EncodeError, InternalError } from "./DataType/error.js";
import { DataTypeSchema } from "./DataType/schema.js";
import * as DataTypeUrl from "./DataTypeUrl.js";
import { AST } from "@effect/schema";

const TypeId: unique symbol = Symbol.for(
  "@blockprotocol/graph/ontology/DataType",
);
export type TypeId = typeof TypeId;

/** @internal */
export const AnnotationId: unique symbol = Symbol.for(
  "@blockprotocol/graph/ontology/DataType/Annotation",
);

export interface DataType<Out, In extends Json.Value = Json.Value>
  extends Equal.Equal,
    Pipeable.Pipeable,
    Inspectable.Inspectable {
  [TypeId]: TypeId;

  readonly id: DataTypeUrl.DataTypeUrl;
  readonly schema: S.Schema<Out, In>;
}

// TODO: cache possibility?!
interface DataTypeImpl<Out, In extends Json.Value = Json.Value>
  extends DataType<Out, In> {}

const DataTypeProto: Omit<DataTypeImpl<unknown>, "id" | "schema"> = {
  [TypeId]: TypeId,

  toJSON(this: DataTypeImpl<unknown>): unknown {
    return {
      _id: "DataType",
      id: this.id,
      schema: this.schema.ast.toJSON(),
    };
  },
  toString(this: DataTypeImpl<unknown>): string {
    return Inspectable.format(this.toJSON());
  },
  [Inspectable.NodeInspectSymbol]() {
    return this.toJSON();
  },

  pipe() {
    // eslint-disable-next-line prefer-rest-params
    Pipeable.pipeArguments(this, arguments);
  },

  [Hash.symbol](this: DataTypeImpl<unknown>) {
    return pipe(
      Hash.hash(TypeId),
      Hash.combine(Hash.hash(this.id)),
      Hash.cached(this),
    );
  },
  [Equal.symbol]<T, E>(this: DataType<T, E>, that: unknown): boolean {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    if (!isDataType(that)) {
      return false;
    }

    return this.id === that.id;
  },
};

const schemaStorage = globalValue(
  Symbol.for("@blockprotocol/graph/ontology/DataType/schemaStorage"),
  () => new Map<DataType<unknown>, DataTypeSchema>(),
);

export function isDataType(value: unknown): value is DataType<unknown> {
  return Predicate.hasProperty(value, TypeId);
}

function makeImpl<Out, In extends Json.Value>(
  id: DataTypeUrl.DataTypeUrl,
  schema: S.Schema<Out, In>,
): DataTypeImpl<Out, In> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const impl = Object.create(DataTypeProto);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  impl.id = id;

  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
  impl.schema = schema.annotations({
    [AnnotationId]: () => impl as DataType<unknown>,
  });

  return impl as DataTypeImpl<Out, In>;
}

function toSchemaImpl<Out, In>(
  schema: S.Schema<Out, In>,
): Effect.Effect<DataTypeSchema, EncodeError> {
  return encodeSchema(schema.ast);
}

export function make<Out, In extends Json.Value>(
  id: DataTypeUrl.DataTypeUrl,
  schema: S.Schema<Out, In>,
): Effect.Effect<DataType<Out, In>, EncodeError> {
  return Effect.gen(function* (_) {
    const impl = makeImpl(id, schema);
    const compiled = yield* _(toSchemaImpl(impl.schema));
    schemaStorage.set(impl as unknown as DataType<unknown>, compiled);
    return impl;
  });
}

export function makeOrThrow<Out, In extends Json.Value>(
  id: DataTypeUrl.DataTypeUrl,
  schema: S.Schema<Out, In>,
): DataType<Out, In> {
  return Effect.runSync(make(id, schema));
}

export function parse<I extends string, Out, In extends Json.Value>(
  id: I,
  schema: S.Schema<Out, In>,
): Effect.Effect<DataType<Out, In>, EncodeError> {
  return pipe(
    DataTypeUrl.parse(id),
    Effect.mapError((error) => EncodeError.invalidUrl(error)),
    Effect.andThen((url) => make(url, schema)),
  );
}

export function parseOrThrow<I extends string, Out, In extends Json.Value>(
  id: I,
  schema: S.Schema<Out, In>,
): DataType<Out, In> {
  return Effect.runSync(parse(id, schema));
}

export function toSchema<Out, In extends Json.Value>(
  dataType: DataType<Out, In>,
): Effect.Effect<DataTypeSchema, EncodeError> {
  const unknownDataType = dataType as unknown as DataType<unknown>;
  if (schemaStorage.has(unknownDataType)) {
    return Effect.succeed(schemaStorage.get(unknownDataType)!);
  }

  return toSchemaImpl(dataType.schema);
}

export function fromSchema(
  schema: DataTypeSchema,
): Effect.Effect<DataType<Json.Value>, DecodeError> {
  return Effect.gen(function* (_) {
    const inner = decodeSchema(schema);

    const dataType = yield* _(
      make(schema.$id, inner),
      Effect.mapError((cause) => DecodeError.encode(cause)),
    );

    return dataType;
  });
}

/** @internal */
export const tryFromAST = (
  ast: AST.AST,
): Effect.Effect<DataType<unknown>, InternalError> =>
  Effect.gen(function* (_) {
    const annotation = AST.getAnnotation(ast, AnnotationId);
    if (Option.isNone(annotation)) {
      return yield* _(InternalError.annotation("missing"));
    }

    if (!Predicate.isFunction(annotation.value)) {
      return yield* _(InternalError.annotation("expected function"));
    }

    const dataType = annotation.value();
    if (!isDataType(dataType)) {
      return yield* _(
        InternalError.annotation("expected function to return `DataType`"),
      );
    }

    return dataType;
  });

/** @internal */
export const getFromAST = (
  ast: AST.AST,
): Effect.Effect<Option.Option<DataType<unknown>>> =>
  pipe(tryFromAST(ast), Effect.option);

export type { DataTypeSchema as Schema };
