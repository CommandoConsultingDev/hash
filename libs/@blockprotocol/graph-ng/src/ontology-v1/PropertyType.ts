import * as S from "@effect/schema/Schema";

import * as DataType from "./DataType";
import * as PropertyTypeUrl from "./PropertyTypeUrl";
import { Brand, Function, Predicate } from "effect";
import * as VersionedUrl from "../VersionedUrl";
import * as Property from "../knowledge/Property";
import * as Json from "../internal/Json";

interface OneOfFrom<T> {
  oneOf: ReadonlyArray<T>;
}

interface OneOf<T, A extends ReadonlyArray<T> = ReadonlyArray<T>> {
  oneOf: A;
}

const OneOf = <To, From, C>(
  oneOf: S.Schema<To, From, C>,
): S.Schema<OneOf<To>, OneOf<From>, C> =>
  S.struct({
    oneOf: S.array(oneOf).pipe(S.minItems(1)),
  });

// TODO: more tuple variants
function makeValueSchemaOneOfPropertyValues<
  A extends ReadonlyArray<PropertyValues>,
>(
  value: OneOf<PropertyValues, A>,
): A extends readonly [infer B extends PropertyValues]
  ? ReturnType<typeof makeValueSchemaPropertyValues<B>>
  : unknown {
  // TODO: actually make it not unknown lol
  return S.union(
    ...value.oneOf.map((item) => makeValueSchemaPropertyValues(item)),
  ) as never;
}

interface Array<T> {
  minItems?: number | undefined;
  maxItems?: number | undefined;

  items: T;
}

const Array = <To, From, C>(
  items: S.Schema<To, From, C>,
): S.Schema<Array<To>, Array<From>, C> =>
  S.struct({
    minItems: S.optional(S.number.pipe(S.int(), S.positive())),
    maxItems: S.optional(S.number.pipe(S.int(), S.positive())),

    items,
  });

type ValueSchemaProperty<T extends PropertyTypeUrl.PropertyTypeUrl> = S.Schema<
  Property.Property<T>,
  Json.Value
>;

function makeValueSchemaProperty<T extends PropertyTypeUrl.PropertyTypeUrl>(
  url: T,
): ValueSchemaProperty<T> {
  return S.transform(
    Json.Value,
    Property.Property,
    (value) => ({
      id: url,
      value,
    }),
    (value) => value.value,
  ) as never;
}

type ValueSchemaArrayPropertyType<T extends PropertyTypeUrl.PropertyTypeUrl> =
  ReturnType<typeof makeValueSchemaArrayProperty<T>>;

function makeValueSchemaArrayProperty<
  T extends PropertyTypeUrl.PropertyTypeUrl,
>(value: Array<T>) {
  return S.array(makeValueSchemaProperty(value.items)).pipe(
    Predicate.isNotUndefined(value.minItems)
      ? S.minItems(value.minItems)
      : Function.identity,
    Predicate.isNotUndefined(value.maxItems)
      ? S.maxItems(value.maxItems)
      : Function.identity,
  );
}

interface PropertyTypeObject
  extends ReadonlyArray<
    PropertyTypeUrl.PropertyTypeUrl | Array<PropertyTypeUrl.PropertyTypeUrl>
  > {}

interface PropertyTypeObjectFrom
  extends ReadonlyArray<
    | S.Schema.From<typeof PropertyTypeUrl.PropertyTypeUrl>
    | Array<S.Schema.From<typeof PropertyTypeUrl.PropertyTypeUrl>>
  > {}

const PropertyTypeObject: S.Schema<PropertyTypeObject, PropertyTypeObjectFrom> =
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  S.array(
    S.suspend(() =>
      S.union(
        PropertyTypeUrl.PropertyTypeUrl,
        Array(PropertyTypeUrl.PropertyTypeUrl),
      ),
    ),
  );

type ValueSchemaPropertyTypeObjectSingleUnknown = S.Schema<
  {
    [key in string]:
      | S.Schema.To<ValueSchemaProperty<PropertyTypeUrl.PropertyTypeUrl>>
      | S.Schema.To<
          ValueSchemaArrayPropertyType<PropertyTypeUrl.PropertyTypeUrl>
        >;
  },
  {
    [key in string]: Json.Value;
  }
>;

type ValueSchemaPropertyTypeObjectSingleProperty<
  T extends PropertyTypeUrl.PropertyTypeUrl,
> = S.Schema<
  Record<"a", S.Schema.To<ValueSchemaProperty<T>>>,
  Record<VersionedUrl.Base<T>, Json.Value>
>;

type ValueSchemaPropertyTypeObjectSingleArray<
  T extends PropertyTypeUrl.PropertyTypeUrl,
> = S.Schema<
  {
    [key in VersionedUrl.Base<T>]: S.Schema.To<ValueSchemaArrayPropertyType<T>>;
  },
  {
    [key in VersionedUrl.Base<T>]: Json.Value;
  }
>;

function makeValueSchemaPropertyTypeObjectSingle<
  U extends
    | PropertyTypeUrl.PropertyTypeUrl
    | Array<PropertyTypeUrl.PropertyTypeUrl>,
>(
  value: U,
): [U] extends [Array<infer T extends PropertyTypeUrl.PropertyTypeUrl>]
  ? ValueSchemaPropertyTypeObjectSingleArray<T>
  : [U] extends [PropertyTypeUrl.PropertyTypeUrl]
  ? ValueSchemaPropertyTypeObjectSingleProperty<U>
  : ValueSchemaPropertyTypeObjectSingleUnknown {
  let valueSchema;
  let valueBase;

  if (Predicate.hasProperty(value, "items")) {
    valueSchema = makeValueSchemaArrayProperty(value);
    valueBase = VersionedUrl.base(value.items);
  } else {
    valueSchema = makeValueSchemaProperty(value);
    valueBase = VersionedUrl.base(value);
  }

  // never here because of generics
  return S.struct({
    [valueBase]: valueSchema,
  }) as never;
}

// TODO: more tuple versions c:
function makeValueSchemaPropertyTypeObject<T extends PropertyTypeObject>(
  value: T,
): T extends readonly [
  infer A extends
    | PropertyTypeUrl.PropertyTypeUrl
    | Array<PropertyTypeUrl.PropertyTypeUrl>,
]
  ? ReturnType<typeof makeValueSchemaPropertyTypeObjectSingle<A>>
  : ValueSchemaPropertyTypeObjectSingleUnknown {
  return value
    .map((value) => {
      return makeValueSchemaPropertyTypeObjectSingle(value);
    })
    .reduce(
      (acc, value) => S.extend(acc, value) as never,
      S.struct({}),
    ) as never;
}

interface ArrayOfPropertyValues<
  T extends PropertyTypeUrl.PropertyTypeUrl = PropertyTypeUrl.PropertyTypeUrl,
> extends Array<OneOf<PropertyValues<T>>> {}
interface ArrayOfPropertyValuesFrom
  extends Array<OneOfFrom<PropertyValuesFrom>> {}

const ArrayOfPropertyValues: S.Schema<
  ArrayOfPropertyValues,
  ArrayOfPropertyValuesFrom
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
> = Array(OneOf(S.suspend(() => PropertyValues)));

type ValueSchemaArrayOfPropertyValuesUnknown = S.Schema<
  readonly S.Schema.To<ValueSchemaPropertyValuesUnknown>[],
  readonly S.Schema.From<ValueSchemaPropertyValuesUnknown>[]
>;

function makeValueSchemaArrayOfPropertyValues<
  T extends PropertyTypeUrl.PropertyTypeUrl,
>(
  value: Array<OneOf<PropertyValues<T>>>,
): ValueSchemaArrayOfPropertyValuesUnknown {
  let values = makeValueSchemaOneOfPropertyValues(value.items);

  // TODO: idk if that's correct here tbh
  return S.array(values as any).pipe(
    Predicate.isNotUndefined(value.minItems)
      ? S.minItems(value.minItems)
      : Function.identity,
    Predicate.isNotUndefined(value.maxItems)
      ? S.maxItems(value.maxItems)
      : Function.identity,
  ) as never;
}

type PropertyValues<
  T extends PropertyTypeUrl.PropertyTypeUrl = PropertyTypeUrl.PropertyTypeUrl,
> = DataType.DataType | PropertyTypeObject | ArrayOfPropertyValues<T>;

type PropertyValuesFrom =
  | S.Schema.From<typeof DataType.DataType>
  | PropertyTypeObjectFrom
  | ArrayOfPropertyValuesFrom;

const PropertyValues: S.Schema<PropertyValues, PropertyValuesFrom> = S.union(
  DataType.DataType,
  PropertyTypeObject,
  ArrayOfPropertyValues,
);

type ValueSchemaPropertyValuesUnknown = S.Schema<
  | S.Schema.To<DataType.ValueSchema<DataType.DataType>>
  | S.Schema.To<ReturnType<typeof makeValueSchemaPropertyTypeObject>>
  | S.Schema.To<ValueSchemaArrayOfPropertyValuesUnknown>,
  Json.Value
>;

function makeValueSchemaPropertyValues<
  P extends PropertyValues = PropertyValues,
>(
  value: P,
): [P] extends [DataType.DataType]
  ? DataType.ValueSchema<P>
  : [P] extends [PropertyTypeObject]
  ? ReturnType<typeof makeValueSchemaPropertyTypeObject<P>>
  : [P] extends [ArrayOfPropertyValues<infer T>]
  ? ReturnType<typeof makeValueSchemaArrayOfPropertyValues<T>>
  : ValueSchemaPropertyValuesUnknown {
  if (Predicate.hasProperty(value, "kind")) {
    return DataType.makeValueSchema(value) as never;
  }

  if (Predicate.hasProperty(value, "items")) {
    return makeValueSchemaArrayOfPropertyValues(value) as never;
  }

  return makeValueSchemaPropertyTypeObject(value) as never;
}

export interface PropertyType<
  A extends ReadonlyArray<PropertyValues> = ReadonlyArray<PropertyValues>,
> extends OneOf<PropertyValues, A> {
  kind: "propertyType";

  id: PropertyTypeUrl.PropertyTypeUrl;
  title: string;
  description?: string | undefined;
}

interface PropertyTypeFrom extends OneOfFrom<PropertyValuesFrom> {
  kind: "propertyType";

  id: S.Schema.From<typeof PropertyTypeUrl.PropertyTypeUrl>;
  title: string;
  description?: string | undefined;
}

export const PropertyType: S.Schema<PropertyType, PropertyTypeFrom> = S.extend(
  S.struct({
    kind: S.literal("propertyType"),

    id: PropertyTypeUrl.PropertyTypeUrl,
    title: S.string,
    description: S.optional(S.string),
  }),

  OneOf(PropertyValues),
);

type ValueSchema<
  A extends ReadonlyArray<PropertyValues> = ReadonlyArray<PropertyValues>,
> = ReturnType<typeof makeValueSchemaOneOfPropertyValues<A>>;

export function makeValueSchema<A extends ReadonlyArray<PropertyValues>>(
  value: PropertyType<A>,
): ValueSchema<A> {
  return makeValueSchemaOneOfPropertyValues(value);
}

//TODO: maybe instead of a full tree, once we hit another data type (via reference or such) we just fallback
// to an opaque value, that is tagged with that type and can be resolved later. That removes any of the annoying
// circular dependencies and such, and makes the validation lighter (although a bit more cumbersome to use) as one
// needs to generate and pass in the schema for every type, but it allows for easier validation and such.
// if we just have: `Property<T>($id, value)` and then `PropertyType.resolve($type, value)` that should work, and we
// can use the same type in the graph!
