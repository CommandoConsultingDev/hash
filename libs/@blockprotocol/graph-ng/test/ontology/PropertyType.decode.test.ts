import * as S from "@effect/schema/Schema";
import { Effect, ReadonlyArray, ReadonlyRecord } from "effect";
import { NoSuchElementException } from "effect/Cause";
import { expect, test } from "vitest";

import * as DataType from "../../src/ontology/DataType.js";
import * as BuiltIn from "../../src/ontology/DataType/BuiltIn.js";
import * as DataTypeUrl from "../../src/ontology/DataTypeUrl.js";
import * as O from "../../src/ontology/OntologySchema.js";
import { OntologyStore } from "../../src/ontology/OntologyStore.js";
import * as PropertyType from "../../src/ontology/PropertyType.js";
import { PropertyTypeSchema } from "../../src/ontology/PropertyType/schema.js";
import * as PropertyTypeUrl from "../../src/ontology/PropertyTypeUrl.js";
import { runError } from "../utils.js";

const TestOntologyStore = (
  propertyTypes: readonly PropertyType.PropertyType<unknown>[],
): OntologyStore<NoSuchElementException> => ({
  dataType: (
    url: DataTypeUrl.DataTypeUrl,
  ): Effect.Effect<DataType.DataType<unknown>, NoSuchElementException, never> =>
    Effect.gen(function* (_) {
      return yield* _(
        BuiltIn as Record<string, Record<string, unknown>>,
        ReadonlyRecord.values,
        ReadonlyArray.flatMap(ReadonlyRecord.values),
        ReadonlyArray.filter(DataType.isDataType),
        ReadonlyArray.findFirst((type) => type.id === url),
      );
    }),
  propertyType: (
    url: PropertyTypeUrl.PropertyTypeUrl,
  ): Effect.Effect<
    PropertyType.PropertyType<unknown>,
    NoSuchElementException,
    never
  > =>
    Effect.gen(function* (_) {
      return yield* _(
        propertyTypes,
        ReadonlyArray.findFirst((type) => type.id === url),
      );
    }),
});

const description = {
  $schema:
    "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
  $id: PropertyTypeUrl.parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/property-type/description/v/1",
  ),
  kind: "propertyType",
  title: "Description",

  oneOf: [
    {
      $ref: BuiltIn.Text.v1.id,
    },
  ],
} satisfies PropertyTypeSchema;

const descriptionType = PropertyType.make(
  description.$id,
  O.dataType(BuiltIn.Text.v1).pipe(S.title("Description")),
).pipe(Effect.runSync);

test("DataType", () => {
  const store = TestOntologyStore([]);

  const propertyType = PropertyType.fromSchema(description, store).pipe(
    Effect.runSync,
  );
  expect(propertyType).toMatchSnapshot();

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(schema("123")).toBe("123");
  expect(() => schema(123)).toThrowErrorMatchingSnapshot();
});

test("DataType: type does not exist", () => {
  const store = TestOntologyStore([]);

  const invalidDescription = {
    ...description,
    oneOf: [
      {
        $ref: "invalid" as DataTypeUrl.DataTypeUrl,
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(invalidDescription, store);
  const error = runError(propertyType);
  expect(error).toMatchSnapshot();
});

const file = {
  $schema:
    "https://blockprotocol.org/types/modules/graph/0.3/schema/property-type",
  $id: PropertyTypeUrl.parseOrThrow(
    "https://blockprotocol.org/@blockprotocol/types/property-type/file/v/1",
  ),
  kind: "propertyType",
  title: "File",

  oneOf: [
    {
      type: "object",
      properties: {
        "https://blockprotocol.org/@blockprotocol/types/property-type/description/":
          {
            $ref: description.$id,
          },
      },
      required: [
        "https://blockprotocol.org/@blockprotocol/types/property-type/description/",
      ],
    },
  ],
} satisfies PropertyTypeSchema;

test("PropertyObject", () => {
  const store = TestOntologyStore([
    descriptionType as unknown as PropertyType.PropertyType<unknown>,
  ]);

  const propertyType = PropertyType.fromSchema(file, store).pipe(
    Effect.runSync,
  );
  expect(propertyType).toMatchSnapshot();

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(
    schema({
      "https://blockprotocol.org/@blockprotocol/types/property-type/description/":
        "123",
    }),
  ).toMatchSnapshot();
  expect(() =>
    schema({
      "https://blockprotocol.org/@blockprotocol/types/property-type/description/": 123,
    }),
  ).toThrowErrorMatchingSnapshot();
});
// TODO: improve errors
test("PropertyObject: type does not exist", () => {
  const store = TestOntologyStore([]);

  const propertyType = PropertyType.fromSchema(file, store);
  const error = runError(propertyType);
  expect(error).toMatchSnapshot();
});
test("PropertyObject: optional", () => {
  const store = TestOntologyStore([
    descriptionType as unknown as PropertyType.PropertyType<unknown>,
  ]);

  const optionalFile = {
    ...file,
    oneOf: [
      {
        ...file.oneOf[0],
        required: [],
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(optionalFile, store).pipe(
    Effect.runSync,
  );

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(schema({})).toMatchSnapshot();
  expect(
    schema({
      "https://blockprotocol.org/@blockprotocol/types/property-type/description/":
        "123",
    }),
  ).toMatchSnapshot();
  expect(() =>
    schema({
      "https://blockprotocol.org/@blockprotocol/types/property-type/description/": 123,
    }),
  ).toThrowErrorMatchingSnapshot();
});
test("ArrayOfPropertyValues", () => {
  const store = TestOntologyStore([]);

  const arrayOfFile = {
    ...file,
    oneOf: [
      {
        type: "array",
        items: {
          oneOf: [
            {
              $ref: BuiltIn.Text.v1.id,
            },
          ],
        },
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(arrayOfFile, store).pipe(
    Effect.runSync,
  );

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(schema(["123"])).toMatchSnapshot();
  expect(() => schema([123])).toThrowErrorMatchingSnapshot();
});
test("ArrayOfPropertyValues: mixed", () => {
  const store = TestOntologyStore([]);

  const arrayOfFile = {
    ...file,
    oneOf: [
      {
        type: "array",
        items: {
          oneOf: [
            {
              $ref: BuiltIn.Text.v1.id,
            },
            {
              $ref: BuiltIn.Number.v1.id,
            },
          ],
        },
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(arrayOfFile, store).pipe(
    Effect.runSync,
  );

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(schema(["123", 123])).toMatchSnapshot();
  expect(() => schema([[]])).toThrowErrorMatchingSnapshot();
});

test("mixed", () => {
  const store = TestOntologyStore([
    descriptionType as unknown as PropertyType.PropertyType<unknown>,
  ]);

  const mixedFile = {
    ...file,
    oneOf: [
      {
        type: "array",
        items: {
          oneOf: [
            {
              $ref: BuiltIn.Text.v1.id,
            },
            {
              type: "object",
              properties: {
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/":
                  {
                    $ref: description.$id,
                  },
              },
              required: [
                "https://blockprotocol.org/@blockprotocol/types/property-type/description/",
              ],
            },
          ],
        },
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(mixedFile, store).pipe(
    Effect.runSync,
  );

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(schema(["123"])).toMatchSnapshot();
  expect(
    schema([
      {
        "https://blockprotocol.org/@blockprotocol/types/property-type/description/":
          "123",
      },
    ]),
  ).toMatchSnapshot();
  expect(() => schema([123])).toThrowErrorMatchingSnapshot();
});

test("recursive", () => {
  const store = TestOntologyStore([]);

  const recursiveFile = {
    ...file,

    oneOf: [
      {
        type: "object",
        properties: {
          "https://blockprotocol.org/@blockprotocol/types/property-type/file/":
            {
              $ref: file.$id,
            },
        },
        required: [],
      },
    ],
  } as PropertyTypeSchema;

  const propertyType = PropertyType.fromSchema(recursiveFile, store).pipe(
    Effect.runSync,
  );

  const schema = S.decodeUnknownSync(propertyType.schema);
  expect(() => schema(123)).toThrowErrorMatchingSnapshot();
  expect(
    schema({
      "https://blockprotocol.org/@blockprotocol/types/property-type/file/": {
        "https://blockprotocol.org/@blockprotocol/types/property-type/file/": {
          "https://blockprotocol.org/@blockprotocol/types/property-type/file/":
            {},
        },
      },
    }),
  ).toMatchSnapshot();
});
