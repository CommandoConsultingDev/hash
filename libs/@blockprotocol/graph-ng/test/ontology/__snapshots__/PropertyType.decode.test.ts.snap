// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArrayOfPropertyValues 1`] = `
[
  "123",
]
`;

exports[`ArrayOfPropertyValues 2`] = `
[Error: File
└─ [0]
   └─ <refinement schema>
      └─ From side refinement failure
         └─ Expected An ordered sequence of characters, actual 123]
`;

exports[`ArrayOfPropertyValues: mixed 1`] = `
[
  "123",
  123,
]
`;

exports[`ArrayOfPropertyValues: mixed 2`] = `
[Error: File
└─ [0]
   └─ <refinement schema> | <refinement schema>
      ├─ Union member
      │  └─ <refinement schema>
      │     └─ From side refinement failure
      │        └─ Expected An ordered sequence of characters, actual []
      └─ Union member
         └─ <refinement schema>
            └─ From side refinement failure
               └─ Expected An arithmetical value (in the Real number system), actual []]
`;

exports[`DataType 1`] = `
{
  "_id": "PropertyType",
  "id": "https://blockprotocol.org/@blockprotocol/types/property-type/description/v/1",
  "schema": {
    "_tag": "Refinement",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/PropertyType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "Description",
    },
    "from": {
      "_tag": "StringKeyword",
      "annotations": {
        "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
        "Symbol(@effect/schema/annotation/Description)": "An ordered sequence of characters",
        "Symbol(@effect/schema/annotation/Title)": "Text",
      },
    },
  },
}
`;

exports[`DataType 2`] = `
[Error: Description
└─ From side refinement failure
   └─ Expected An ordered sequence of characters, actual 123]
`;

exports[`DataType: type does not exist 1`] = `
{
  "_id": "Cause",
  "_tag": "Fail",
  "failure": {
    "_tag": "@blockprotocol/graph/PropertyType/DecodeError",
    "reason": {
      "_tag": "FetchDataType",
      "cause": [NoSuchElementException],
    },
  },
}
`;

exports[`PropertyObject 1`] = `
{
  "_id": "PropertyType",
  "id": "https://blockprotocol.org/@blockprotocol/types/property-type/file/v/1",
  "schema": {
    "_tag": "TypeLiteral",
    "annotations": {
      "Symbol(@blockprotocol/graph/ontology/PropertyType/Annotation)": [Function],
      "Symbol(@effect/schema/annotation/Title)": "File",
    },
    "indexSignatures": [],
    "propertySignatures": [
      {
        "annotations": {},
        "isOptional": false,
        "isReadonly": true,
        "name": "https://blockprotocol.org/@blockprotocol/types/property-type/description/",
        "type": {
          "_tag": "Refinement",
          "annotations": {
            "Symbol(@blockprotocol/graph/ontology/PropertyType/Annotation)": [Function],
            "Symbol(@effect/schema/annotation/Title)": "Description",
          },
          "from": {
            "_tag": "StringKeyword",
            "annotations": {
              "Symbol(@blockprotocol/graph/ontology/DataType/Annotation)": [Function],
              "Symbol(@effect/schema/annotation/Description)": "An ordered sequence of characters",
              "Symbol(@effect/schema/annotation/Title)": "Text",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`PropertyObject 2`] = `
{
  "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "123",
}
`;

exports[`PropertyObject 3`] = `
[Error: File
└─ ["https://blockprotocol.org/@blockprotocol/types/property-type/description/"]
   └─ Description
      └─ From side refinement failure
         └─ Expected An ordered sequence of characters, actual 123]
`;

exports[`PropertyObject: optional 1`] = `{}`;

exports[`PropertyObject: optional 2`] = `
{
  "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "123",
}
`;

exports[`PropertyObject: optional 3`] = `
[Error: File
└─ ["https://blockprotocol.org/@blockprotocol/types/property-type/description/"]
   └─ Description | undefined
      ├─ Union member
      │  └─ Description
      │     └─ From side refinement failure
      │        └─ Expected An ordered sequence of characters, actual 123
      └─ Union member
         └─ Expected undefined, actual 123]
`;

exports[`PropertyObject: type does not exist 1`] = `
{
  "_id": "Cause",
  "_tag": "Fail",
  "failure": {
    "_tag": "@blockprotocol/graph/PropertyType/DecodeError",
    "reason": {
      "_tag": "FetchPropertyType",
      "cause": [NoSuchElementException],
    },
  },
}
`;

exports[`mixed 1`] = `
[
  "123",
]
`;

exports[`mixed 2`] = `
[
  {
    "https://blockprotocol.org/@blockprotocol/types/property-type/description/": "123",
  },
]
`;

exports[`mixed 3`] = `
[Error: File
└─ [0]
   └─ <refinement schema> | { https://blockprotocol.org/@blockprotocol/types/property-type/description/: Description }
      ├─ Union member
      │  └─ <refinement schema>
      │     └─ From side refinement failure
      │        └─ Expected An ordered sequence of characters, actual 123
      └─ Union member
         └─ Expected { https://blockprotocol.org/@blockprotocol/types/property-type/description/: Description }, actual 123]
`;

exports[`recursive 1`] = `[Error: Expected File, actual 123]`;

exports[`recursive 2`] = `
{
  "https://blockprotocol.org/@blockprotocol/types/property-type/file/": {
    "https://blockprotocol.org/@blockprotocol/types/property-type/file/": {
      "https://blockprotocol.org/@blockprotocol/types/property-type/file/": {},
    },
  },
}
`;
